src/kzg10/subproductdomain.rs

```diff
@@ -0,0 +1,424 @@
+
+        left.evaluate(&r_0, u_0, t_0);
+        right.evaluate(&r_1, u_1, t_1);
+    }
+    /// Fast interpolate over this subproduct tree
+    pub fn interpolate(&self, u: &[F], v: &[F]) -> Poly<F> {
+        let mut lagrange_coeff = self.inverse_lagrange_coefficients(u);
+
+        for (s_i, v_i) in lagrange_coeff.iter_mut().zip(v.iter()) {
+            *s_i = s_i.inverse().unwrap() * *v_i;
+        }
+
+        self.linear_combine(u, &lagrange_coeff)
+    }
+    /// Fast compute lagrange coefficients over this subproduct tree
+    pub fn inverse_lagrange_coefficients(&self, u: &[F]) -> Vec<F> {
+        //assert u.len() == degree of s.m
+        if u.len() == 1 {
+            return vec![F::one()];
+        }
+        let mut evals = vec![F::zero(); u.len()];
+        let m_prime = derivative::<F>(&self.m);
+        self.evaluate(&m_prime, u, &mut evals);
+        evals
+    }
+    /// GG Algorithm 10.9
+    /// Fast linear combination of moduli over this subproduct tree
+    /// On input c = { c_0, ..., c_{n-1} }
+    /// output sum_i (c_i * m) / (x- u_i)
+    /// Takes O(M(n) log n) field operations
+    pub fn linear_combine(&self, u: &[F], c: &[F]) -> Poly<F> {
+        if u.len() == 1 {
+            // Output c_0 * (x-u_0) / (x-u_0) = c_0
+            return Poly::<F> { coeffs: vec![c[0]] };
+        }
+        let n = u.len() / 2;
+        let (u_0, u_1) = u.split_at(n);
+        let (c_0, c_1) = c.split_at(n);
+
+        // Linear combination of moduli over both halves of the subproduct tree
+        let left = self.left.as_ref().unwrap();
+        let right = self.right.as_ref().unwrap();
+        let r_0 = left.linear_combine(u_0, c_0);
+        let r_1 = right.linear_combine(u_1, c_1);
+
+        // r_0 = sum_{i in left} c_i m_left / (x-u_i)
+        // so m_right * r_0 = sum_{i in left} c_i (m_left*m_right) / (x-u_i) = sum_{i in left} c_i m / (x-u_i)
+        &(&right.m * &r_0) + &(&left.m * &r_1)
+    }
+}
+/// compute the derivative of polynomial f
+pub fn derivative<F: FftField>(f: &Poly<F>) -> Poly<F> {
+    let mut coeffs = Vec::with_capacity(f.coeffs().len() - 1);
+    for (i, c) in f.coeffs.iter().enumerate().skip(1) {
+        coeffs.push(F::from(i as u64) * c);
+    }
+    Poly::<F> { coeffs }
+}
+
+/// Build a vector representation of the (n x n) circulant matrix of polynomial f
+/// Based on the blog post:
+/// https://alinush.github.io/2020/03/19/multiplying-a-vector-by-a-toeplitz-matrix.html
+pub fn build_circulant<F: FftField>(f: &Poly<F>, size: usize) -> Vec<F> {
+    let mut circulant = vec![F::zero(); 2 * size];
+    let coeffs = f.coeffs();
+    if size == coeffs.len() - 1 {
+        circulant[0] = *coeffs.last().unwrap();
+        circulant[size] = *coeffs.last().unwrap();
+        circulant[size + 1..size + 1 + coeffs.len() - 2]
+            .copy_from_slice(&coeffs[1..coeffs.len() - 1]);
+    } else {
+        circulant[size + 1..size + 1 + coeffs.len() - 1].copy_from_slice(&coeffs[1..]);
```

Comment:
Unfortunately there is a bug somewhere in here which prevents arbitrary circulant matrices/toeplitz matrices from being constructed (it is not triggered by the KZG opening, but is triggered by other tests where the diagonal of the toeplitz matrix is all zeros)

Reply:
What's the bug that you face?

Reply:
it doesn't give the correct result for `circulant` in the "else" case (e.g. zeros on the diagonal of the Toeplitz matrix). The "else" is never triggered by the fast KZG open so all of those tests pass.

I suspect it's just some indexing error, off by one or something, but I haven't debugged it yet.